### DISCUSSION TITLE
How to inject a file into a bundle within a custom plugin?

### DISCUSSION THREAD

**Discussion by tlhunter** [webpack/webpack#16530] [Q&A]
I'm working on adding Webpack support for a popular APM product. One of the requirements is that the APM knows the versions of instrumented libraries (redis, pg, mysql) at runtime. Usually this means calling some semblance of `require('db/package.json')`. This would be fine if the package itself requires its own `package.json` file as it would be added to the bundle automatically. But most packages don't do this.

What I'm trying to do is add the appropriate package.json file to the bundle at runtime. Based on the method name, `complication.addModule()` sounds promising, but after reading documentation and source code I'm still at a loss. I'm hoping to do something like this:

```javascript
const contents = require('redis/package.json'); // should the contents be read?

// should I instantiate a new webpack.Module? sounds correct but can't find it documented anywhere
const packageJson = new webpack.Module('what is module type? appears undocumented');

packageJson.something(); // are there methods to set file contents? or path to file which gets read?

compilation.addModule(packageJson); // it seems that addModule expects a Module instance
```

**Update**: I realized I didn't mention this in the post but I'm specifically trying to create a Webpack plugin, not modify an application.

**Comment by AllanOricil** [comment]
Did you try the json-loader plugin? If yes, why can't it solve your problem?

**Reply by tlhunter** [reply]
I did not try to use the `json-loader` plugin as it appears to be incompatible with Webpack v4+. When I looked at the `json-loader` source it appears to only translate JSON files which are encountered by Webpack when building the module hierarchy based on `require()` calls, which is now built into Webpack v2+.\

I'm hoping for a way to manually / proactively add a subset of JSON files where Webpack won't find a correlating `require()` call.

**Reply by alexander-akait** [reply]
You can use `new URL("./path/to/package.json", import.meta.url)` or register `asset` module for require extensions/files
